File: LeetCode/108/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    TreeNode* construct(vector<int> arr, int low, int high) {

        if(low > high) {
            return nullptr;
        }

        int mid = low + (high - low) / 2;
        
        TreeNode* node = new TreeNode(arr[mid]);
        node->left = construct(arr, low, mid - 1);
        node->right = construct(arr, mid + 1, high);

        return node;
    }

    TreeNode* sortedArrayToBST(vector<int>& arr) {
        int n = arr.size();
        return construct(arr, 0, n - 1);
    }
};

File: LeetCode/108/main.py
"""
DIFFICULTY : easy
TAGS : array, tree, bst, divide and conquer, binary tree
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def helper(self, arr, left, right):
        mid = left + (right - left) // 2
        if left > right:
            return
        root = TreeNode(arr[mid])
        root.left = self.helper(arr, left, mid - 1)
        root.right = self.helper(arr, mid + 1, right)
        return root

    def sortedArrayToBST(self, arr: List[int]) -> Optional[TreeNode]:
        return self.helper(arr, 0, len(arr) - 1)
